bot-components/GitLab_subscriptions.ml

```diff
@@ -0,0 +1,117 @@
+open Base
+open Cohttp
+open Utils
+open Yojson.Basic.Util
+
+let extract_commit json =
+  let open Yojson.Basic.Util in
+  let commit_json = json |> member "commit" in
+  let message = commit_json |> member "message" |> to_string in
+  if string_match ~regexp:"Bot merge .* into \\(.*\\)" message then
+    Str.matched_group 1 message
+  else
+    (* In the case of build webhooks, the id is a number and the sha is the
+       reference of the commit, while in the case of pipeline hooks only id
+       is present and represents the sha. *)
+    ( match commit_json |> member "sha" with
+    | `Null ->
+        commit_json |> member "id"
+    | sha ->
+        sha )
+    |> to_string
+
+let pr_from_branch branch =
+  if string_match ~regexp:"^pr-\\([0-9]*\\)$" branch then
+    (Some (Str.matched_group 1 branch |> Int.of_string), "pull request")
+  else (None, "branch")
+
+type job_info =
+  { build_status: string
+  ; build_id: int
+  ; build_name: string
+  ; commit: string
+  ; branch: string
+  ; repo_url: string
+  ; project_id: int
+  ; failure_reason: string option
+  ; allow_fail: bool option }
+
+type pipeline_info =
+  {state: string; id: int; commit: string; branch: string; project_path: string}
+
+let job_info_of_json json =
+  let open Yojson.Basic.Util in
+  let build_status = json |> member "build_status" |> to_string in
+  let build_id = json |> member "build_id" |> to_int in
+  let build_name = json |> member "build_name" |> to_string in
+  let project_id = json |> member "project_id" |> to_int in
+  let commit = json |> extract_commit in
+  let branch = json |> member "ref" |> to_string in
+  let repo_url = json |> member "repository" |> member "url" |> to_string in
+  let failure_reason, allow_fail =
+    if String.equal build_status "failed" then
+      ( json |> member "build_failure_reason" |> to_string |> Option.some
+      , json |> member "build_allow_failure" |> to_bool |> Option.some )
+    else (None, None)
+  in
+  { build_status
+  ; build_id
+  ; build_name
+  ; commit
+  ; branch
+  ; repo_url
+  ; project_id
+  ; failure_reason
+  ; allow_fail }
+
+let pipeline_info_of_json json =
+  let open Yojson.Basic.Util in
+  let pipeline_json = json |> member "object_attributes" in
+  let state = pipeline_json |> member "status" |> to_string in
+  let id = pipeline_json |> member "id" |> to_int in
+  let commit = json |> extract_commit in
+  let branch =
+    json |> member "object_attributes" |> member "ref" |> to_string
+  in
+  let project_path =
+    json |> member "project" |> member "path_with_namespace" |> to_string
+  in
+  {state; id; commit; branch; project_path}
+
+let gitlab_event ~event json =
+  match event with
+  | "Job Hook" ->
+      Ok (`JobEvent (job_info_of_json json))
+  | "Pipeline Hook" ->
+      Ok (`PipelineEvent (pipeline_info_of_json json))
+  | e ->
+      Ok (`UnsupportedEvent e)
```

Comment:
I suggest that you create a datatype for GitLab events, as was done in `GitHub_subscriptions` for GitHub events.